[Hannah Loh] iP

hhlh27 opened the pull request:
# DukePro

>“Your mind is for having ideas, not holding them.” – David Allen (source)

DukePro frees your mind of having to remember things you need to do. It's:

- **text-based**
- **easy to learn**
- **~~FAST~~  SUPER FAST** to use

## How to Use

All you need to do is:

1. Download it from [here](#).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉.

And it is **FREE!**
This site was built using [GitHub Pages](https://pages.github.com/).
## Features

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

hhlh27 committed:
Add greeting and exit messages

hhlh27 committed:
Add user input handling

hhlh27 committed:
Add add and list functions

hhlh27 committed:
Add Task.java class

hhlh27 committed:
Update Task.java

hhlh27 committed:
Add mark as done feature

hhlh27 committed:
Separate features into different methods

hhlh27 committed:
Add Deadline.java

hhlh27 committed:
Add Event.java

hhlh27 committed:
Add ToDo.java

hhlh27 committed:
Add todo and deadline tasks feature

hhlh27 committed:
Add deadline task feature

hhlh27 committed:
Add error handling

hhlh27 committed:
Add delete task function

hhlh27 committed:
Add enum for command type

hhlh27 committed:
Edit files for automated testing

hhlh27 committed:
Add save tasks feature

hhlh27 committed:
Add getter methods for Deadline and Task classes

hhlh27 committed:
Merge branch 'branch-Level-7'

hhlh27 committed:
Modify Deadline class to use LocalDateTime

hhlh27 committed:
Modify Event class to handle start and end times

hhlh27 committed:
Add dates and times handling feature

hhlh27 committed:
Merge branch 'branch-Level-8'

hhlh27 committed:
Add UI.java class

hhlh27 committed:
Add TaskList.java class

hhlh27 committed:
Add Command.java

hhlh27 committed:
Add ui.java class

hhlh27 committed:
Add Parser.java class

hhlh27 committed:
Add TaskList.java class

hhlh27 committed:
Add Storage.java class

hhlh27 committed:
Add command classes

hhlh27 committed:
Update DazAI.java class

hhlh27 committed:
Merge branch 'branch-A-MoreOOP'

hhlh27 committed:
Put classes into package

hhlh27 committed:
Merge branch 'add-gradle-support'

hhlh27 committed:
Add gradle

hhlh27 committed:
Add JUnit tests

hhlh27 committed:
Add JavaDoc comments

hhlh27 committed:
Add java doc

hhlh27 committed:
Modify code to comply with coding standard

hhlh27 committed:
Merge tag 'A-CodingStandard'

hhlh27 committed:
Add find function

hhlh27 committed:
Merge branch 'branch-Level-9'

hhlh27 committed:
Merge branch 'branch-A-JavaDoc'

hhlh27 committed:
Add checkstyle files

hhlh27 committed:
Merge branch 'Branch-A-CheckStyle'

hhlh27 committed:
Change DazAI to DazAi

hhlh27 committed:
Add GUI

hhlh27 committed:
Modify GUI

hhlh27 committed:
Merge branch 'branch-Level-10'

hhlh27 committed:
Add Varargs

hhlh27 committed:
Merge branch 'branch-A-Varargs'

gohlucas commented on src/main/java/dazai/AddCommand.java:
```diff
@@ -0,0 +1,86 @@
+
+    private final String type;
+    private final String description;
+    private final String dateTime1; // Used for deadlines and events
+    private final String dateTime2; // Used for events only
```
Could these variables have a more specific name to indicate their purpose?

gohlucas commented on src/main/java/dazai/AddCommand.java:
```diff
@@ -0,0 +1,86 @@
+     * @throws DazAiException If the task type or parameters are invalid.
+     */
+    private Task createTask() throws DazAiException {
+        switch (type) {
+            case "event":
+                if (dateTime1 == null || dateTime2 == null) {
+                    throw new DazAiException("Invalid event format! Use: event <desc> /from yyyy-MM-dd HHmm /to yyyy-MM-dd HHmm");
+                }
+                return new Event(description, dateTime1, dateTime2);
+            case "deadline":
+                if (dateTime1 == null) {
+                    throw new DazAiException("Invalid deadline format! Use: deadline <desc> /by yyyy-MM-dd HHmm");
+                }
+                return new Deadline(description, dateTime1);
+            case "todo":
+                if (description.isEmpty()) {
+                    throw new DazAiException("The description of a todo cannot be empty.");
+                }
+                return new ToDo(description);
+            default:
+                throw new DazAiException("Invalid task type! Use: event, deadline, or todo.");
+        }
```
```suggestion
        switch (type) {
        case "event":
            if (dateTime1 == null || dateTime2 == null) {
                throw new DazAiException("Invalid event format! Use: event <desc> /from yyyy-MM-dd HHmm /to yyyy-MM-dd HHmm");
            }
            return new Event(description, dateTime1, dateTime2);
        case "deadline":
            if (dateTime1 == null) {
                throw new DazAiException("Invalid deadline format! Use: deadline <desc> /by yyyy-MM-dd HHmm");
            }
            return new Deadline(description, dateTime1);
        case "todo":
            if (description.isEmpty()) {
                throw new DazAiException("The description of a todo cannot be empty.");
            }
            return new ToDo(description);
        default:
            throw new DazAiException("Invalid task type! Use: event, deadline, or todo.");
        }
```

gohlucas replied:
I believe this is more in line with the Java Coding Standards and is something that is not intuitive initially as I have made this mistake as well

gohlucas commented on src/main/java/dazai/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Deadline extends Task {
+    private LocalDateTime by;
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    private static final DateTimeFormatter OUTPUT_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
```
I like how you saved it as a static final variable for easy reference to the DateTimeFormat Standard for manipulation of user input

gohlucas commented on src/main/java/dazai/MainWindow.java:
```diff
@@ -0,0 +1,57 @@
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
```
Perhaps there is an error in the naming of this method

gohlucas commented on src/main/java/dazai/Ui.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Displays the welcome message when the program starts.
+     */
+    public void showWelcome() {
+        String logo = "  ____                    ___    _____ \n"
+                + " |  _ \\  __ _  _____     / _\\   |__  _|\n"
+                + " | | | |/ _` | |_  |    / /  \\    | | \n"
+                + " | |_| | (_| |  / /_   /  --- \\   | | \n"
+                + " |____/ \\__,_| |__,_| //       \\\\|____|  \n";
+
+        // Print greeting message
+        System.out.println("____________________________________________________________");
+        System.out.println(" Hello from\n" + logo);
+        System.out.println(" Hello! I'm DazAi");
+        System.out.println(" What can I do for you?");
```
I love the effort and enthusiasm to recreate a logo unique to your project!!

gohlucas commented on src/main/java/dazai/Parser.java:
```diff
@@ -0,0 +1,139 @@
+     */
+    private static Command parseDeadline(String[] words) throws DazAiException {
+        if (words.length < 2) {
+            throw new DazAiException("Invalid deadline format! Use: deadline <desc> /by yyyy-MM-dd HHmm");
```
Considering that it is used quite extensively in this file, would extracting the exception message into a separate string be neater?

gohlucas reviewed (commented):
LGTM. I like how intuitve the naming of the methods are!

kararei commented on src/main/java/dazai/AddCommand.java:
```diff
@@ -0,0 +1,86 @@
+     * @throws DazAiException If the task type or parameters are invalid.
+     */
+    private Task createTask() throws DazAiException {
+        switch (type) {
+            case "event":
+                if (dateTime1 == null || dateTime2 == null) {
+                    throw new DazAiException("Invalid event format! Use: event <desc> /from yyyy-MM-dd HHmm /to yyyy-MM-dd HHmm");
+                }
+                return new Event(description, dateTime1, dateTime2);
+            case "deadline":
+                if (dateTime1 == null) {
+                    throw new DazAiException("Invalid deadline format! Use: deadline <desc> /by yyyy-MM-dd HHmm");
+                }
+                return new Deadline(description, dateTime1);
+            case "todo":
+                if (description.isEmpty()) {
+                    throw new DazAiException("The description of a todo cannot be empty.");
+                }
+                return new ToDo(description);
+            default:
+                throw new DazAiException("Invalid task type! Use: event, deadline, or todo.");
+        }
```
```suggestion
        switch (type) {
        case "event":
            if (dateTime1 == null || dateTime2 == null) {
                throw new DazAiException("Invalid event format! Use: event <desc> /from yyyy-MM-dd HHmm /to yyyy-MM-dd HHmm");
            }
            return new Event(description, dateTime1, dateTime2);
        case "deadline":
            if (dateTime1 == null) {
                throw new DazAiException("Invalid deadline format! Use: deadline <desc> /by yyyy-MM-dd HHmm");
            }
            return new Deadline(description, dateTime1);
        case "todo":
            if (description.isEmpty()) {
                throw new DazAiException("The description of a todo cannot be empty.");
            }
            return new ToDo(description);
        default:
            throw new DazAiException("Invalid task type! Use: event, deadline, or todo.");
    }
```

kararei commented on src/main/java/dazai/Storage.java:
```diff
@@ -0,0 +1,94 @@
+                    default:
+                        System.out.println("Skipping unknown task type: " + line);
+                        continue;
+                }
```
```suggestion
                switch (taskType) {
                case "T":
                    task = new ToDo(description);
                    break;
                case "D":
                    if (taskData.length < 4) {
                        System.out.println("Skipping invalid deadline task: " + line);
                        continue;
                    }
                    task = new Deadline(description, taskData[3].trim());
                    break;
                case "E":
                    if (taskData.length < 5) {
                        System.out.println("Skipping invalid event task: " + line);
                        continue;
                    }
                    task = new Event(description, taskData[3].trim(), taskData[4].trim());
                    break;
                default:
                    System.out.println("Skipping unknown task type: " + line);
                    continue;
            }
```

kararei replied:
Add proper indentation for switch blocks according to the java naming conventions.

kararei commented on src/main/java/dazai/Storage.java:
```diff
@@ -0,0 +1,94 @@
+            for (Task task : tasks.getAllTasks()) {
+                String taskType = task instanceof ToDo ? "T" :
+                        task instanceof Deadline ? "D" :
+                                task instanceof Event ? "E" : "?";
```
Instead of the if-else approach used here, you could use a more object-oriented approach. By having a method getTaskType() in each of the "ToDo", "Deadline" and "Event" file, it would use polymorphism for better code clarity.

e.g. 
public class ToDo extends Task {

    public String getTaskType() {
        return "T";
    }

kararei commented on src/main/java/dazai/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+     * @param description The task description.
+     * @param by The deadline in the format "yyyy-MM-dd HHmm".
+     */
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = LocalDateTime.parse(by, INPUT_FORMAT);
```
```suggestion
    public Deadline(String description, String deadline) {
        super(description);
        this.deadline = LocalDateTime.parse(deadline, INPUT_FORMAT);
```

kararei replied:
Would be clearer to name the parameter "deadline" instead of "by". Also follows naming conventions of using nouns for things and verbs for actions.

kararei commented on src/main/java/dazai/Event.java:
```diff
@@ -0,0 +1,55 @@
+public class Event extends Task {
+    private LocalDateTime from;
+    private LocalDateTime to;
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    private static final DateTimeFormatter OUTPUT_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
```
I like the way you used String Literals to specify and finalise Input and Output formats

kararei reviewed (commented):
lgtm!
I like the OOP used and overall organisation

hhlh27 committed:
Add assertions

hhlh27 committed:
AddCommand class: Correct switch-case formatting and rename datetime variables

The switch-case structure in the AddCommand class was not formatted in the
preferred style. It is essential to follow the coding style for consistency
and readability. Additionally, the variable names for the datetime
fields were not descriptive enough.

Let's reformat the switch-case structure by:
* Removing unnecessary indentation for case clauses
* Adding clear fallthrough comments

Let's rename the datetime variables to improve clarity:
* dateTime1-> fromDateTime
* dateTime2 -> toDateTime

This change improves the overall readability of the code and ensures
consistency with project coding conventions.

hhlh27 committed:
FindCommand class: Add validation for keyword input

The current FindCommand class did not enforce proper validation for keyword inputs.
This could result in empty inputs being keyed in.

Let's introduce validation methods for keyword input to ensure that keyword is not empty or null

By enforcing validation, we can improve the robustness of the system and prevent
empty keywrd search

hhlh27 committed:
Modify code to improve code quality

hhlh27 committed:
Merge pull request #1 from hhlh27/branch-A-Assertions

Add assertions

hhlh27 committed:
Add streams

hhlh27 committed:
Merge pull request #2 from hhlh27/branch-A-CodeQuality

Branch a code quality

hhlh27 committed:
Merge branch 'master' into branch-A-Streams

hhlh27 committed:
Merge pull request #3 from hhlh27/branch-A-Streams

Add streams

hhlh27 committed:
Add tagging function

hhlh27 committed:
Merge branch 'branch-A-Extension'

hhlh27 committed:
Modify codes to imrpove code quality

hhlh27 committed:
Storage class: Add Javadocs to improve documentation quality
The Storage class is currently missing detailed Javadoc
comments for its public methods.
The lack of documentation makes it difficult for other
developers to understand the purpose and usage of the
methods.

Javadoc comments have been added to:

loadTasks: Explains the loading of tasks from the file,
the expected format, and potential exceptions.
saveTasks: Describes how tasks are serialized
and saved to the storage file.
Let's ensure the comments are formatted properly
 and follow project guidelines,
enhancing code maintainability and readability.

No changes were made to the logic of the class,
focusing solely on documentation improvements.

hhlh27 committed:
FindCommand class: Add Javadocs for improved documentation
quality

The FindCommand class lacked proper Javadoc comments,
making it harder to understand its behavior and usage.

Adding Javadocs improves code readability
and provides clear guidance on:
- The class’s purpose and behavior.
- Constructor parameter requirements
and potential exceptions.
- The execute method’s functionality,
including parameter usage and return details.

This change enhances maintainability and ensures compliance
with the project’s documentation standards.

Let's add Javadocs to provide better documentation
for the FindCommand class,
ensuring clear communication of its intended functionality.

hhlh27 committed:
Update README.md

hhlh27 committed:
Update README.md

hhlh27 committed:
Add user images

hhlh27 committed:
Fix main window display

hhlh27 committed:
Update Ui.png

hhlh27 committed:
Revert "Update Ui.png"

This reverts commit 4b18ca18c6fe31d2e5e105667ab30bcaa67d1e94.

hhlh27 committed:
Merge branch 'master' of https://github.com/hhlh27/ip

hhlh27 committed:
Add Ui.png

hhlh27 committed:
Edit code formatting

hhlh27 committed:
Fix bug in deleteTask function

hhlh27 committed:
Edit code formatting

hhlh27 committed:
Modify build.gradle file

hhlh27 committed:
Update build.gradle file

hhlh27 committed:
Update README.md

hhlh27 committed:
Update README.md
